tests/test_images/test_utils.py

```diff
@@ -0,0 +1,320 @@
+            np.arange(2, 4).reshape(1, -1) * np.arange(2, 5).reshape(-1, 1),
+            None,
+        ),
+        (
+            np.arange(1, 11).reshape(1, -1) * np.arange(1, 11).reshape(-1, 1),
+            None,
+            0.2,
+            np.asarray((4.5, 4.5)),
+            np.arange(2, 10).reshape(1, -1) * np.arange(2, 10).reshape(-1, 1),
+            None,
+        ),
+        (
+            np.arange(1, 11).reshape(1, -1) * np.arange(1, 11).reshape(-1, 1),
+            None,
+            0.2,
+            None,
+            np.arange(2, 10).reshape(1, -1) * np.arange(2, 10).reshape(-1, 1),
+            None,
+        ),
+        (
+            np.arange(1, 11).reshape(1, -1) * np.arange(1, 11).reshape(-1, 1),
+            None,
+            0.4,
+            np.asarray((3.5, 6.5)),
+            np.arange(5, 11).reshape(1, -1) * np.arange(2, 8).reshape(-1, 1),
+            None,
+        ),
+        (
+            np.arange(2).reshape(1, -1) * np.arange(2).reshape(-1, 1),
+            None,
+            2,
+            None,
+            None,
+            InvalidValueError,
+        ),
+        (
+            np.arange(2).reshape(1, -1) * np.arange(2).reshape(-1, 1),
+            None,
+            -1,
+            None,
+            None,
+            InvalidValueError,
+        ),
+        (
+            np.arange(2).reshape(1, -1) * np.arange(2).reshape(-1, 1),
+            None,
+            0.1,
+            np.asarray((3, 1)),
+            None,
+            OutOfBoundsError,
+        ),
+    ],
+    ids=[
+        "unchanged",
+        "shape no offset",
+        "shape stack",
+        "shape offset",
+        "shape too large",
+        "shifted out of bounds",
+        "shape overrides crop",
+        "crop no offset",
+        "crop none offset",
+        "crop offset",
+        "crop >= 1",
+        "crop < 0",
+        "crop shifted out of bounds",
+    ],
+)
+def test_crop(array, target_shape, crop, centre, expected_output, exception):
+    kwargs = {
+        "array": array,
+        "shape": target_shape,
+        "crop": crop,
+        "centre": centre,
+    }
+    if exception is not None:
+        with pytest.raises(exception):
+            utils.apply_crop(**kwargs)
+    else:
+        result = utils.apply_crop(**kwargs)
+        result_2d_shape = np.asarray(result.shape[-2::])
+        if target_shape is not None:
+            assert_array_equal(
+                result_2d_shape, target_shape, err_msg="expected shape does not match"
+            )
+        elif crop != 0:
+            expected_shape = np.round(
+                np.asarray(array.shape[-2::]) * (1.0 - crop)
+            ).astype(np.uint16)
+            assert_array_equal(
+                result_2d_shape, expected_shape, err_msg="expected shape does not match"
+            )
+        assert_array_equal(
+            result, expected_output, err_msg="result doesn't match expected output"
+        )
+
+
+def test_complex_to_interleaved_float(): ...
```

Comment:
Unfinished tests?

Reply:
Ah yeah, removed them as they wouldn't fit in the PR anyway